doc/source/configuration.rst
.. _configuration:

Configuration
=============

Nodepool reads its secure configuration from ``/etc/nodepool/secure.conf``
by default. The secure file is a standard ini config file, with
one section for database, and another section for the jenkins
secrets for each target::

  [database]
  dburi={dburi}

  [jenkins "{target_name}"]
  user={user}
  apikey={apikey}
  credentials={credentials}
  url={url}

Following settings are available::

**required**

  ``dburi``
  Indicates the URI for the database connection.  See the `SQLAlchemy
  documentation
  <http://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_
  for the syntax.  Example::

    dburi='mysql+pymysql://nodepool@localhost/nodepool'

**optional**

  While it is possible to run Nodepool without any Jenkins targets,
  if Jenkins is used, the `target_name` and `url` are required. The
  `user`, `apikey` and `credentials` also may be needed depending on
  the Jenkins security settings.

  ``target_name``
  Name of the jenkins target. It needs to match with a target
  specified in nodepool.yaml, in order to retrieve its settings.

  ``url``
  Url to the Jenkins REST API.

  ``user``
  Jenkins username.

  ``apikey``
  API key generated by Jenkins (not the user password).

  ``credentials``
  If provided, Nodepool will configure the Jenkins slave to use the Jenkins
  credential identified by that ID, otherwise it will use the username and
  ssh keys configured in the image.

Nodepool reads its configuration from ``/etc/nodepool/nodepool.yaml``
by default.  The configuration file follows the standard YAML syntax
with a number of sections defined with top level keys.  For example, a
full configuration file may have the ``labels``, ``providers``, and
``targets`` sections. If building images using diskimage-builder, the
``diskimages`` section is also required::

  labels:
    ...
  diskimages:
    ...
  providers:
    ...
  targets:
    ...

The following sections are available.  All are required unless
otherwise indicated.

script-dir
----------
When creating an image to use when launching new nodes, Nodepool will
run a script that is expected to prepare the machine before the
snapshot image is created.  The ``script-dir`` parameter indicates a
directory that holds all of the scripts needed to accomplish this.
Nodepool will copy the entire directory to the machine before invoking
the appropriate script for the image being created.

Example::

  script-dir: /path/to/script/dir

.. _elements-dir:

elements-dir
------------

If an image is configured to use diskimage-builder and glance to locally
create and upload images, then a collection of diskimage-builder elements
must be present. The ``elements-dir`` parameter indicates a directory
that holds one or more elements.

Example::

  elements-dir: /path/to/elements/dir

images-dir
----------

When we generate images using diskimage-builder they need to be
written to somewhere. The ``images-dir`` parameter is the place to
write them.

Example::

  images-dir: /path/to/images/dir

paths
-----
This section is optional.

Nodepool relies on external applications such as disk-image-builder. Typically
these are found by the environment PATH however in certain situations we want
to configure these to a specific location. This allows configuration to
override a path to the application.

Example::

  disk-image-create: /path/to/disk-image-create

cron
----
This section is optional.

Nodepool runs several periodic tasks.  The ``image-update`` task
creates a new image for each of the defined images, typically used to
keep the data cached on the images up to date.  The ``cleanup`` task
deletes old images and servers which may have encountered errors
during their initial deletion.  The ``check`` task attempts to log
into each node that is waiting to be used to make sure that it is
still operational.  The following illustrates how to change the
schedule for these tasks and also indicates their default values::

  cron:
    image-update: '14 2 * * *'
    cleanup: '27 */6 * * *'
    check: '*/15 * * * *'

zmq-publishers
--------------
Lists the ZeroMQ endpoints for the Jenkins masters.  Nodepool uses
this to receive real-time notification that jobs are running on nodes
or are complete and nodes may be deleted.  Example::

  zmq-publishers:
    - tcp://jenkins1.example.com:8888
    - tcp://jenkins2.example.com:8888

gearman-servers
---------------
Lists the Zuul Gearman servers that should be consulted for real-time
demand.  Nodepool will use information from these servers to determine
if additional nodes should be created to satisfy current demand.
Example::

  gearman-servers:
    - host: zuul.example.com
      port: 4730

The ``port`` key is optional (default: 4730).

zookeeper-servers
-----------------
Lists the ZooKeeper servers uses for coordinating information between
nodepool workers. Example::

  zookeeper-servers:
    - host: zk1.example.com
      port: 2181
      chroot: /nodepool

The ``port`` key is optional (default: 2181).

The ``chroot`` key, used for interpreting ZooKeeper paths relative to
the supplied root path, is also optional and has no default.

.. _labels:

labels
------

Defines the types of nodes that should be created.  Maps node types to
the images that are used to back them and the providers that are used
to supply them.  Jobs should be written to run on nodes of a certain
label (so targets such as Jenkins don't need to know about what
providers or images are used to create them).  Example::

  labels:
    - name: my-precise
      image: precise
      min-ready: 2
      providers:
        - name: provider1
        - name: provider2
    - name: multi-precise
      image: precise
      subnodes: 2
      min-ready: 2
      ready-script: setup_multinode.sh
      providers:
        - name: provider1

**required**

  ``name``
    Unique name used to tie jobs to those instances.

  ``image``
    Refers to providers images, see :ref:`images`.

  ``providers`` (list)
    Required if any nodes should actually be created (e.g., the label is not
    currently disabled, see ``min-ready`` below).

**optional**

  ``min-ready`` (default: 2)
    Minimum instances that should be in a ready state. Set to -1 to have the
    label considered disabled. ``min-ready`` is best-effort based on available
    capacity and is not a guaranteed allocation.

  ``subnodes``
    Used to configure multi-node support.  If a `subnodes` key is supplied to
    an image, it indicates that the specified number of additional nodes of the
    same image type should be created and associated with each node for that
    image.

    Only one node from each such group will be added to the target, the
    subnodes are expected to communicate directly with each other.  In the
    example above, for each Precise node added to the target system, two
    additional nodes will be created and associated with it.

  ``ready-script``
    A script to be used to perform any last minute changes to a node after it
    has been launched but before it is put in the READY state to receive jobs.
    For more information, see :ref:`scripts`.

.. _diskimages:

diskimages
----------

Lists the images that are going to be built using diskimage-builder.
Image keyword defined on labels section will be mapped to the
images listed on diskimages. If an entry matching the image is found
this will be built using diskimage-builder and the settings found
on this configuration. If no matching image is found, image
will be built using the provider snapshot approach::

  diskimages:
  - name: devstack-precise
    elements:
      - ubuntu
      - vm
      - puppet
      - nodepool-base
      - node-devstack
    release: precise
    env-vars:
        DIB_DISTRIBUTION_MIRROR: http://archive.ubuntu.com
        DIB_IMAGE_CACHE: /opt/dib_cache


**required**

  ``name``
    Identifier to reference the disk image in :ref:`images` and :ref:`labels`.

**optional**

  ``release``
    Specifies the distro to be used as a base image to build the image using
    diskimage-builder.

  ``elements`` (list)
    Enumerates all the elements that will be included when building the image,
    and will point to the :ref:`elements-dir` path referenced in the same
    config file.

  ``env-vars`` (dict)
    Arbitrary environment variables that will be available in the spawned
    diskimage-builder child process.

.. _provider:

provider
---------

Lists the OpenStack cloud providers Nodepool should use.  Within each
provider, the Nodepool image types are also defined (see
:ref:`images` for details).  Example::

  providers:
    - name: provider1
      cloud: example
      region-name: 'region1'
      max-servers: 96
      rate: 1.0
      availability-zones:
        - az1
      boot-timeout: 120
      launch-timeout: 900
      template-hostname: 'template-{image.name}-{timestamp}'
      pool: 'public'
      ipv6-preferred: False
      networks:
        - name: 'some-network-name'
          public: True
      images:
        - name: trusty
          base-image: 'Trusty'
          min-ram: 8192
          name-filter: 'something to match'
          setup: prepare_node.sh
          username: jenkins
          user-home: '/home/jenkins'
          private-key: /var/lib/jenkins/.ssh/id_rsa
          meta:
              key: value
              key2: value
        - name: precise
          base-image: 'Precise'
          min-ram: 8192
          setup: prepare_node.sh
          username: jenkins
          user-home: '/home/jenkins'
          private-key: /var/lib/jenkins/.ssh/id_rsa
        - name: devstack-trusty
          min-ram: 30720
          diskimage: devstack-trusty
          username: jenkins
          private-key: /home/nodepool/.ssh/id_rsa
    - name: provider2
      username: 'username'
      password: 'password'
      auth-url: 'http://auth.provider2.example.com/'
      project-name: 'project'
      service-type: 'compute'
      service-name: 'compute'
      region-name: 'region1'
      max-servers: 96
      rate: 1.0
      template-hostname: '{image.name}-{timestamp}-nodepool-template'
      images:
        - name: precise
          base-image: 'Fake Precise'
          min-ram: 8192
          setup: prepare_node.sh
          username: jenkins
          user-home: '/home/jenkins'
          private-key: /var/lib/jenkins/.ssh/id_rsa
          meta:
              key: value
              key2: value

**cloud configuration***

**preferred**

  ``cloud``
  There are two methods supported for configuring cloud entries. The preferred
  method is to create an ``~/.config/openstack/clouds.yaml`` file containing
  your cloud configuration information. Then, use ``cloud`` to refer to a
  named entry in that file.

  More information about the contents of `clouds.yaml` can be found in
  `the os-client-config documentation <http://docs.openstack.org/developer/os-client-config/>`_.

**compatablity**

  For backwards compatibility reasons, you can also include
  portions of the cloud configuration directly in ``nodepool.yaml``. Not all
  of the options settable via ``clouds.yaml`` are available.

  ``username``

  ``password``

  ``project-id`` OR ``project-name``
    Some clouds may refer to the ``project-id`` as ``tenant-id``.
    Some clouds may refer to the ``project-name`` as ``tenant-name``.

  ``auth-url``
    Keystone URL.

  ``image-type``
    Specifies the image type supported by this provider.  The disk images built
    by diskimage-builder will output an image for each ``image-type`` specified
    by a provider using that particular diskimage.

    By default, ``image-type`` is set to the value returned from
    ``os-client-config`` and can be omitted in most cases.

**required**

  ``name``

  ``max-servers``
    Maximum number of servers spawnable on this provider.

**optional**

  ``availability-zones`` (list)
    Without it nodepool will rely on nova to schedule an availability zone.

    If it is provided the value should be a list of availability zone names.
    Nodepool will select one at random and provide that to nova. This should
    give a good distribution of availability zones being used. If you need more
    control of the distribution you can use multiple logical providers each
    providing a different list of availabiltiy zones.

  ``boot-timeout``
    Once an instance is active, how long to try connecting to the
    image via SSH.  If the timeout is exceeded, the node launch is
    aborted and the instance deleted.

    In seconds. Default 60.

  ``launch-timeout``

    The time to wait from issuing the command to create a new instance
    until that instance is reported as "active".  If the timeout is
    exceeded, the node launch is aborted and the instance deleted.

    In seconds. Default 3600.

  ``keypair``
    Default None

  ``networks`` (dict)
    Specify custom Neutron networks that get attached to each
    node. Specify the ``name`` of the network (a string) and if the
    network routes to the Internet, set the boolean ``public`` to
    true.

  ``ipv6-preferred``
    If it is set to True, nodepool will try to find ipv6 in public net first
    as the ip address for ssh connection to build snapshot images and create
    jenkins slave definition. If ipv6 is not found or the key is not
    specified or set to False, ipv4 address will be used.

  ``pool``
    Specify a floating ip pool in cases where the 'public' pool is unavailable
    or undesirable.

  ``api-timeout`` (compatability)
    Timeout for the OpenStack API calls client in seconds. Prefer setting
    this in `clouds.yaml`

  ``service-type`` (compatability)
    Prefer setting this in `clouds.yaml`.

  ``service-name`` (compatability)
    Prefer setting this in `clouds.yaml`.

  ``region-name``

  ``template-hostname``
    Hostname template to use for the spawned instance.
    Default ``template-{image.name}-{timestamp}``

  ``rate``
    In seconds. Default 1.0.

  ``clean-floating-ips``
    If it is set to True, nodepool will assume it is the only user of the
    OpenStack project and will attempt to clean unattached floating ips that
    may have leaked around restarts.

.. _images:

images
~~~~~~

Example::

  images:
    - name: precise
      base-image: 'Precise'
      min-ram: 8192
      name-filter: 'something to match'
      setup: prepare_node.sh
      username: jenkins
      private-key: /var/lib/jenkins/.ssh/id_rsa
      meta:
          key: value
          key2: value

**required**

  ``name``
    Identifier to refer this image from :ref:`labels` and :ref:`provider`
    sections.

    If the resulting images from different providers ``base-image`` should be
    equivalent, give them the same name; e.g. if one provider has a ``Fedora
    20`` image and another has an equivalent ``Fedora 20 (Heisenbug)`` image,
    they should use a common ``name``.  Otherwise select a unique ``name``.

  ``base-image``
    UUID or string-name of the image to boot as specified by the provider.

  ``min-ram``
    Determine the flavor of ``base-image`` to use (e.g. ``m1.medium``,
    ``m1.large``, etc).  The smallest flavor that meets the ``min-ram``
    requirements will be chosen. To further filter by flavor name, see optional
    ``name-filter`` below.

**optional**

  ``name-filter``
    Additional filter complementing ``min-ram``, will be required to match on
    the flavor-name (e.g. Rackspace offer a "Performance" flavour; setting
    `name-filter` to ``Performance`` will ensure the chosen flavor also
    contains this string as well as meeting `min-ram` requirements).

  ``setup``
     Script to run to prepare the instance.

     Used only when not building images using diskimage-builder, in that case
     settings defined in the ``diskimages`` section will be used instead. See
     :ref:`scripts` for setup script details.

  ``diskimages``
     See :ref:`diskimages`.

  ``username``
    Nodepool expects that user to exist after running the script indicated by
    ``setup``. Default ``jenkins``

  ``private-key``
    Default ``/var/lib/jenkins/.ssh/id_rsa``

  ``config-drive`` (boolean)
    Whether config drive should be used for the image.

  ``meta`` (dict)
    Arbitrary key/value metadata to store for this server using the Nova
    metadata service. A maximum of five entries is allowed, and both keys and
    values must be 255 characters or less.

.. _targets:

targets
-------

Lists the Jenkins masters to which Nodepool should attach nodes after
they are created.  Nodes of each label will be evenly distributed
across all of the targets which are on-line::

  targets:
    - name: jenkins1
      hostname: '{label.name}-{provider.name}-{node_id}'
      subnode-hostname: '{label.name}-{provider.name}-{node_id}-{subnode_id}'
    - name: jenkins2
      hostname: '{label.name}-{provider.name}-{node_id}'
      subnode-hostname: '{label.name}-{provider.name}-{node_id}-{subnode_id}'

**required**

  ``name``
  Identifier for the system an instance is attached to.

**optional**

  ``hostname``
    Default ``{label.name}-{provider.name}-{node_id}``

  ``subnode-hostname``
    Default ``{label.name}-{provider.name}-{node_id}-{subnode_id}``

  ``rate``
    In seconds. Default 1.0

  ``jenkins`` (dict)

    ``test-job`` (optional)
      Setting this would cause a newly created instance to be in a TEST state.
      The job name given will then be executed with the node name as a
      parameter.

      If the job succeeds, move the node into READY state and relabel it with
      the appropriate label (from the image name).

      If it fails, immediately delete the node.

      If the job never runs, the node will eventually be cleaned up by the
      periodic cleanup task.


nodepool/config.py
#!/usr/bin/env python

# Copyright (C) 2011-2013 OpenStack Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
#
# See the License for the specific language governing permissions and
# limitations under the License.

import os_client_config
from six.moves import configparser as ConfigParser
import yaml

import fakeprovider


class ConfigValue(object):
    def __eq__(self, other):
        if isinstance(other, ConfigValue):
            if other.__dict__ == self.__dict__:
                return True
        return False


class Config(ConfigValue):
    pass


class Provider(ConfigValue):
    def __eq__(self, other):
        if (other.cloud_config != self.cloud_config or
            other.max_servers != self.max_servers or
            other.pool != self.pool or
            other.image_type != self.image_type or
            other.rate != self.rate or
            other.api_timeout != self.api_timeout or
            other.boot_timeout != self.boot_timeout or
            other.launch_timeout != self.launch_timeout or
            other.networks != self.networks or
            other.ipv6_preferred != self.ipv6_preferred or
            other.clean_floating_ips != self.clean_floating_ips or
            other.azs != self.azs):
            return False
        new_images = other.images
        old_images = self.images
        # Check if images have been added or removed
        if set(new_images.keys()) != set(old_images.keys()):
            return False
        # check if existing images have been updated
        for k in new_images:
            if (new_images[k].base_image != old_images[k].base_image or
                new_images[k].min_ram != old_images[k].min_ram or
                new_images[k].name_filter != old_images[k].name_filter or
                new_images[k].setup != old_images[k].setup or
                new_images[k].username != old_images[k].username or
                new_images[k].user_home != old_images[k].user_home or
                new_images[k].diskimage != old_images[k].diskimage or
                new_images[k].private_key != old_images[k].private_key or
                new_images[k].meta != old_images[k].meta or
                new_images[k].config_drive != old_images[k].config_drive):
                return False
        return True

    def __ne__(self, other):
        return not self.__eq__(other)


class ProviderImage(ConfigValue):
    pass


class Target(ConfigValue):
    pass


class Label(ConfigValue):
    pass


class LabelProvider(ConfigValue):
    pass


class Cron(ConfigValue):
    pass


class ZMQPublisher(ConfigValue):
    pass


class GearmanServer(ConfigValue):
    pass


class ZooKeeperServer(ConfigValue):
    pass


class DiskImage(ConfigValue):
    pass


class Network(ConfigValue):
    pass


def loadConfig(config_path):
    config = yaml.load(open(config_path))
    cloud_config = os_client_config.OpenStackConfig()

    newconfig = Config()
    newconfig.db = None
    newconfig.dburi = None
    newconfig.providers = {}
    newconfig.targets = {}
    newconfig.labels = {}
    newconfig.scriptdir = config.get('script-dir')
    newconfig.elementsdir = config.get('elements-dir')
    newconfig.imagesdir = config.get('images-dir')
    newconfig.paths = config.get('paths', {})
    newconfig.dburi = None
    newconfig.provider_managers = {}
    newconfig.jenkins_managers = {}
    newconfig.zmq_publishers = {}
    newconfig.gearman_servers = {}
    newconfig.zookeeper_servers = {}
    newconfig.diskimages = {}
    newconfig.crons = {}

    for name, default in [
        ('image-update', '14 2 * * *'),
        ('cleanup', '* * * * *'),
        ('check', '*/15 * * * *'),
        ]:
        c = Cron()
        c.name = name
        newconfig.crons[c.name] = c
        c.job = None
        c.timespec = config.get('cron', {}).get(name, default)

    for addr in config.get('zmq-publishers', []):
        z = ZMQPublisher()
        z.name = addr
        z.listener = None
        newconfig.zmq_publishers[z.name] = z

    for server in config.get('gearman-servers', []):
        g = GearmanServer()
        g.host = server['host']
        g.port = server.get('port', 4730)
        g.name = g.host + '_' + str(g.port)
        newconfig.gearman_servers[g.name] = g

    for server in config.get('zookeeper-servers', []):
        z = ZooKeeperServer()
        z.host = server['host']
        z.port = server.get('port', 2181)
        z.chroot = server.get('chroot', '')
        z.name = z.host + '_' + str(z.port)
        newconfig.zookeeper_servers[z.name] = z

    for provider in config.get('providers', []):
        p = Provider()
        p.name = provider['name']
        newconfig.providers[p.name] = p

        cloud_kwargs = _cloudKwargsFromProvider(provider)
        p.cloud_config = _get_one_cloud(cloud_config, cloud_kwargs)
        p.region_name = provider.get('region-name')
        p.max_servers = provider['max-servers']
        p.keypair = provider.get('keypair', None)
        p.pool = provider.get('pool', None)
        p.rate = provider.get('rate', 1.0)
        p.api_timeout = provider.get('api-timeout')
        p.boot_timeout = provider.get('boot-timeout', 60)
        p.launch_timeout = provider.get('launch-timeout', 3600)
        p.networks = []
        for network in provider.get('networks', []):
            n = Network()
            p.networks.append(n)
            if 'net-id' in network:
                n.id = network['net-id']
                n.name = None
            elif 'net-label' in network:
                n.name = network['net-label']
                n.id = None
            else:
                n.name = network.get('name')
                n.id = None
            n.public = network.get('public', False)
        p.ipv6_preferred = provider.get('ipv6-preferred')
        p.clean_floating_ips = provider.get('clean-floating-ips')
        p.azs = provider.get('availability-zones')
        p.template_hostname = provider.get(
            'template-hostname',
            'template-{image.name}-{timestamp}'
        )
        p.image_type = provider.get(
            'image-type', p.cloud_config.config['image_format'])
        p.images = {}
        for image in provider['images']:
            i = ProviderImage()
            i.name = image['name']
            p.images[i.name] = i
            i.base_image = image.get('base-image', None)
            i.min_ram = image['min-ram']
            i.name_filter = image.get('name-filter', None)
            i.setup = image.get('setup', None)
            i.diskimage = image.get('diskimage', None)
            i.username = image.get('username', 'jenkins')
            i.user_home = image.get('user-home', '/home/jenkins')
            i.private_key = image.get('private-key',
                                      '/var/lib/jenkins/.ssh/id_rsa')
            i.config_drive = image.get('config-drive', None)

            # note this does "double-duty" -- for
            # SnapshotImageUpdater the meta-data dict is passed to
            # nova when the snapshot image is created.  For
            # DiskImageUpdater, this dict is expanded and used as
            # custom properties when the image is uploaded.
            i.meta = image.get('meta', {})
            # 5 elements, and no key or value can be > 255 chars
            # per Nova API rules
            if i.meta:
                if len(i.meta) > 5 or \
                   any([len(k) > 255 or len(v) > 255
                        for k, v in i.meta.iteritems()]):
                    # soft-fail
                    #self.log.error("Invalid metadata for %s; ignored"
                    #               % i.name)
                    i.meta = {}

    if 'diskimages' in config:
        for diskimage in config['diskimages']:
            d = DiskImage()
            d.name = diskimage['name']
            newconfig.diskimages[d.name] = d
            if 'elements' in diskimage:
                d.elements = u' '.join(diskimage['elements'])
            else:
                d.elements = ''
            # must be a string, as it's passed as env-var to
            # d-i-b, but might be untyped in the yaml and
            # interpreted as a number (e.g. "21" for fedora)
            d.release = str(diskimage.get('release', ''))
            d.env_vars = diskimage.get('env-vars', {})
            if not isinstance(d.env_vars, dict):
                #self.log.error("%s: ignoring env-vars; "
                #               "should be a dict" % d.name)
                d.env_vars = {}
            d.image_types = set()
        # Do this after providers to build the image-types
        for provider in newconfig.providers.values():
            for image in provider.images.values():
                if (image.diskimage and
                    image.diskimage in newconfig.diskimages):
                    diskimage = newconfig.diskimages[image.diskimage]
                    diskimage.image_types.add(provider.image_type)

    for label in config.get('labels', []):
        l = Label()
        l.name = label['name']
        newconfig.labels[l.name] = l
        l.image = label['image']
        l.min_ready = label.get('min-ready', 2)
        l.subnodes = label.get('subnodes', 0)
        l.ready_script = label.get('ready-script')
        l.providers = {}
        for provider in label['providers']:
            p = LabelProvider()
            p.name = provider['name']
            l.providers[p.name] = p

    for target in config.get('targets', []):
        t = Target()
        t.name = target['name']
        newconfig.targets[t.name] = t
        jenkins = target.get('jenkins', {})
        t.online = True
        t.rate = target.get('rate', 1.0)
        t.jenkins_test_job = jenkins.get('test-job')
        t.jenkins_url = None
        t.jenkins_user = None
        t.jenkins_apikey = None
        t.jenkins_credentials_id = None

        t.assign_via_gearman = target.get('assign-via-gearman', False)

        t.hostname = target.get(
            'hostname',
            '{label.name}-{provider.name}-{node_id}'
        )
        t.subnode_hostname = target.get(
            'subnode-hostname',
            '{label.name}-{provider.name}-{node_id}-{subnode_id}'
        )

    # A set of image names that are in use by labels, to be
    # used by the image update methods to determine whether
    # a given image needs to be updated.
    newconfig.images_in_use = set()
    for label in newconfig.labels.values():
        if label.min_ready >= 0:
            newconfig.images_in_use.add(label.image)

    return newconfig


def loadSecureConfig(config, secure_config_path):
    secure = ConfigParser.ConfigParser()
    secure.readfp(open(secure_config_path))

    config.dburi = secure.get('database', 'dburi')

    for target in config.targets.values():
        section_name = 'jenkins "%s"' % target.name
        if secure.has_section(section_name):
            target.jenkins_url = secure.get(section_name, 'url')
            target.jenkins_user = secure.get(section_name, 'user')
            target.jenkins_apikey = secure.get(section_name, 'apikey')

        try:
            target.jenkins_credentials_id = secure.get(
                section_name, 'credentials')
        except:
            pass


def _cloudKwargsFromProvider(provider):
    cloud_kwargs = {}
    for arg in ['region-name', 'api-timeout', 'cloud']:
        if arg in provider:
            cloud_kwargs[arg] = provider[arg]

    # These are named from back when we only talked to Nova. They're
    # actually compute service related
    if 'service-type' in provider:
        cloud_kwargs['compute-service-type'] = provider['service-type']
    if 'service-name' in provider:
        cloud_kwargs['compute-service-name'] = provider['service-name']

    auth_kwargs = {}
    for auth_key in (
            'username', 'password', 'auth-url', 'project-id', 'project-name'):
        if auth_key in provider:
            auth_kwargs[auth_key] = provider[auth_key]

    cloud_kwargs['auth'] = auth_kwargs
    return cloud_kwargs


def _get_one_cloud(cloud_config, cloud_kwargs):
    '''This is a function to allow for overriding it in tests.'''
    if cloud_kwargs.get('auth', {}).get('auth-url', '') == 'fake':
        return fakeprovider.fake_get_one_cloud(cloud_config, cloud_kwargs)
    return cloud_config.get_one_cloud(**cloud_kwargs)
